### ISSUE TITLE
gnome-shell session is terminated when systemd-logind is restarted

### ISSUE THREAD

**Issue by keszybz** [systemd/systemd#17308]
**systemd version the issue has been seen with**
 > all, incl. recent git master

**Used distribution**
 > Fedora 33.

This is a long-standing annoyance. In the past we couldn't restart various services without losing state. But since the introduction of fd storage in pid1 all services should be restartable. We fixed systemd-journald a long time ago. In principle systemd-logind should be restartable too, but this doesn't work at all with gnome-shell.

I don't know the cause of the issue. Maybe it's in logind, maybe in gnome-shell, maybe somewhere else. Let's use this ticket to track this. I think there was some special code in gnome-shell (or was it gdm?) to terminate upon logind restart... but I can't find it now.

/cc @benzea

**Comment by keszybz** [comment]
```
[20194.777386] gnome-shell[3219]: Could not open device /dev/input/event2: GDBus.Error:org.freedesktop.login1.NotInControl: You are not in control of this session
[20194.781656] gnome-shell[3219]: Could not open device /dev/input/event9: GDBus.Error:org.freedesktop.login1.NotInControl: You are not in control of this session
[20194.785882] gnome-shell[3219]: Could not open device /dev/input/event0: GDBus.Error:org.freedesktop.login1.NotInControl: You are not in control of this session
[20194.788734] gnome-shell[3219]: Could not open device /dev/input/event1: GDBus.Error:org.freedesktop.login1.NotInControl: You are not in control of this session
[20194.792806] gnome-shell[3219]: Could not open device /dev/input/event6: GDBus.Error:org.freedesktop.login1.NotInControl: You are not in control of this session
```
So it seems that logind removes some permissions too eagerly.

Full log after `systemctl restart systemd-logind` attached.

[log-after-logind-restart.txt](https://github.com/systemd/systemd/files/5359006/log-after-logind-restart.txt)

**Comment by benzea** [comment]
Yeah … I have been wondering and was annoyed by it myself, but to be honest, did not bother to investigate it more so far.

Interesting. What is supposed to happen when logind restarts? Revoking the permissions makes sense, but it also means that there is no active session and that one is in complete desync from the active VT. Will the session first be deactivated by logind during shutdown?

**Comment by keszybz** [comment]
> What is supposed to happen when logind restarts?

Nothing! We should be able to restart logind at any time to allow upgrades and bug fixes.

**Comment by keszybz** [comment]
Also, there's a compounding bug that parts of the old session remain after the crash, and it's not possible to log in again. That's a separate messy bug, though most likely not in logind.

**Comment by benzea** [comment]
I am kind of thinking that the session is still running mostly. And it just never realises that it is becoming active again.

**Comment by poettering** [comment]
logind saves/stores all session data in /run and stores all fds in PID1's fd store. So in theory restarting it should not change stuff (except that in the time window between the stop and the start logind of course won't notice things, won't be on the bus and so on).

A while back we fixed some issues where we didn't serialize things properly, but to my knowledge this should all work again.

I wonder if gnome-shell/gnome-session listen to the bus name disappearing and then do bad stuff. ideally they'd just stick around and continue whatever they do as if nothing happened. Of course if they call into logind while logind is off the bus they would need to handle things gracefully.

**Comment by benzea** [comment]
It tries to save everything. But just stopping logind results in the `CONTROLLER=` information to be lost from the session information. gnome-shell does not seem to be doing anything wrong really, it gets the notification that it is active, tries to do `TakeDevice`, but fails because logind forgot that it is the controller.

Looks to me like:
 * manager is unref'ed
 * manager destroy session
 * `session_free` calls `session_drop_controller`
 * `session_drop_controller` calls `session_save`
 * Session is saved without the controller information

I guess all we need to do is prevent `session_drop_controller` (and any later call in `session_free`) from saving the session.

**Comment by poettering** [comment]
oh, that could be. it looks like a logind bug then indeed!

**Comment by keszybz** [comment]
It seems systemd-logind needs adjusting to not close various fds and leave resources in place. I prepped some patches to do that, but that is clearly not enough to prevent gnome from going down. In fact, just closing the bus connection in logind is enough to bring the graphical session down.

```console
$ sudo gdb -p $(pidof systemd-logind)
(gdb) up 3
(gdb) p sd_bus_close(m->bus)
$3 = void
```
This causes:
```
Nov 09 20:30:49 fedora gnome-shell[10062]: Could not release device (13,64): Cannot invoke method; proxy is for the well-known name org.freedesktop.login1 without an owner, and proxy was constructed with the G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START flag
Nov 09 20:30:49 fedora gnome-shell[10062]: Could not release device (13,66): Cannot invoke method; proxy is for the well-known name org.freedesktop.login1 without an owner, and proxy was constructed with the G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START flag
Nov 09 20:30:49 fedora gnome-shell[10062]: Could not release device (13,65): Cannot invoke method; proxy is for the well-known name org.freedesktop.login1 without an owner, and proxy was constructed with the G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START flag
Nov 09 20:30:49 fedora gnome-shell[10062]: Could not release device (13,67): Cannot invoke method; proxy is for the well-known name org.freedesktop.login1 without an owner, and proxy was constructed with the G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START flag
```
That message is from mutter code: https://sources.debian.org/src/mutter/3.38.1-2/src/backends/native/meta-launcher.c/?hl=377#L377. 

So it seems like this will require some changes in gnome-shell/mutter too. @benzea, would you be willing to look into this from that side? I know very little about mutter code, so it wouldn't be effective for me to try to debug this.

https://github.com/systemd/systemd/pull/17558 has my patches so far.

**Comment by keszybz** [comment]
I'm also unsetting the milestone since this seems to be a complicated issue that we're not likely to fix quickly.

**Comment by benzea** [comment]
Hmm, something odd must be happening when systemd-logind disappears from the bus. Getting errors when it tries to release the device is one thing (that seems somewhat reasonable), but I guess it shouldn't even be doing that in this case.

i.e. gnome-shell should be able to ignore the fact that systemd-logind has disappeared. As I understand it, it should be able to assume that everything remains the same while systemd-logind is gone and things will be sync'ed properly once systemd-logind comes back.

**Comment by benzea** [comment]
And yeah, can look into it :)

**Comment by keszybz** [comment]
> Getting errors when it tries to release the device is one thing (that seems somewhat reasonable), but I guess it shouldn't even be doing that in this case.

Yeah, trying to drop the device over dbus, in response to the dbus connection going away, has little chance of success ;)
FWIW, it's pretty clear that it's dropping all the fds it got from logind:
```
Nov 09 20:28:35 fedora systemd[1]: Stopping User Login Management...
Nov 09 20:28:35 fedora systemd-logind[9153]: Bus n/a: changing state RUNNING → CLOSED
Nov 09 20:28:35 fedora gnome-shell[9292]: Could not release device (13,64): Cannot invoke method; proxy is for the well-known name org.freedesktop.login1 without an owner, and proxy was constructed with the G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START flag
Nov 09 20:28:35 fedora systemd-logind[9153]: Leaving fd in store (28, session-c3-device-13-67)
Nov 09 20:28:35 fedora systemd-logind[9153]: Leaving fd in store (5, session-c3-device-226-0)
Nov 09 20:28:35 fedora gnome-shell[9292]: Could not release device (13,66): Cannot invoke method; proxy is for the well-known name org.freedesktop.login1 without an owner, and proxy was constructed with the G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START flag
Nov 09 20:28:35 fedora gnome-shell[9292]: Could not release device (13,65): Cannot invoke method; proxy is for the well-known name org.freedesktop.login1 without an owner, and proxy was constructed with the G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START flag
Nov 09 20:28:35 fedora gnome-shell[9292]: Could not release device (13,67): Cannot invoke method; proxy is for the well-known name org.freedesktop.login1 without an owner, and proxy was constructed with the G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START flag
Nov 09 20:28:35 fedora systemd-logind[9153]: Leaving fd in store (23, session-c3-device-13-65)
Nov 09 20:28:35 fedora systemd-logind[9153]: Leaving fd in store (6, session-c3-device-13-64)
Nov 09 20:28:35 fedora systemd-logind[9153]: Leaving fd in store (7, session-c3-device-13-66)
```
(Messages are interleaved because stuff happens in parallel. But actually the response from gnome-shell is only in response to the bus close.)

> And yeah, can look into it :)

That'd be great, thanks!

**Comment by benzea** [comment]
OK, started on an MR for mutter: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1556

I have not yet tested it because that isn't the most convenient thing to test without risking loosing the session. But, pretty sure that fixes the fact that we try to release the devices.

(Oh, fun, didn't know that gitlab would show the reference here thanks to the mirrors.)

**Comment by benzea** [comment]
@keszybz so, I just tried things with my patch (by killing systemd-logind rather than shutting it down cleanly, to avoid it corrupting the stored data).

Things seemed to work nicely in general, but:
 1. I got switched to vt1 immediately (logind started again right away)
 2. My tty seems to have gotten input. i.e. everything worked, but I hit `ctrl+c` and apparently that did job-control foo on my session leader process and killed my session

**Comment by benzea** [comment]
But, disconnecting systemd-logind from the bus didn't affect my session in any way.

**Comment by keszybz** [comment]
Thanks for the mutter patches. With that patch merged, I can stop systemd-logind seemingly without any immediate effect on my session. The session is killed if *restart* logind. But things are clearly moving in the right direction.

**Comment by julienfalque** [comment]
Since I updated Ubuntu from 20.10 to 21.04, when I try to suspend the system, Gnome seems to crash and sends me back to login screen. If I try to login again, I only get  a black screen with a few `^@` characters printed. If I instead try to shutdown, GDM warns that a session is still active.

System logs do contain several occurences of
```
gnome-shell[3374]: Could not release device (13,67): Cannot invoke method; proxy is for the well-known name org.freedesktop.login1 without an owner, and proxy was constructed with the G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START flag
```
at the time of crash so I think this is related to this issue. How can I be sure it is? How can I help solving it?

**Comment by benzea** [comment]
> How can I be sure it is? How can I help solving it?

I doubt it is, there is no reason to have a logind restart. What happens for you is that gnome-session is not properly shutting down and you need to wait 1.5 minutes for GDM to notice that the session has died.

**Comment by mbiebl** [comment]
I've filed this a while ago
https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/1881

Being able to restart logind after a package upgrade would be great.
We track this issue downstream at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=919509

**Comment by septatrix** [comment]
Is this isolated to GNOME or are KDE, wlroots etc also affected?

**Comment by Vladimir-csp** [comment]
I can confirm that standalone compositors are affected, tested with Hyprland, Labwc, Wayfire, Sway, Weston.

But the strange thing is: the session is NOT terminated. Its tty is put into foreground instead.

Compositor (along with all its clients) continues to run "normally" "somewhere" in the background as if nothing happened, but becomes totally inaccessible. It even reports monitors as available/active if queried via IPC.

But switching to compositor's tty just brings up tty, `loginctl activate ...` does the same.  Session type is downgraded from `wayland` to `tty`.

Before logind restart:

```
244 - username (1000)
  Since: Sun 2025-09-28 13:22:17 MSK; 2min 21s ago
  State: active
 Leader: 44798 (login)
   Seat: seat0; vc1
    TTY: tty1
 Remote: no
Service: login
   Type: wayland
  Class: user
   Idle: no
   Unit: session-244.scope
         ├─44798 "login -- username"
         ├─45239 systemctl --user start --wait wayland-wm@labwc.desktop.service
         └─45535 waitpid -e 45660

Sep 28 13:22:17 hostname systemd[1]: Started session-244.scope - Session 244 of User username.
Sep 28 13:22:19 hostname uwsm[45239]: Default compositor ID unchanged: labwc.desktop.
Sep 28 13:22:19 hostname uwsm[45239]: Selected compositor ID: labwc.desktop
                                                Command Line: labwc
                                            Plugin/binary ID: labwc
                                       Initial Desktop Names: labwc:wlroots
                                                        Name: labwc
                                                 Description: A wayland stacking compositor
Sep 28 13:22:19 hostname uwsm[45239]: Units unchanged.
Sep 28 13:22:19 hostname uwsm[45239]: Starting labwc.desktop and waiting while it is running...
Sep 28 13:22:20 hostname uwsm[45535]: waitpid: Holding login session 244 open until PID 45660 exits
```

After logind restart:

```
244 - username (1000)
  Since: Sun 2025-09-28 13:22:17 MSK; 2min 46s ago
  State: online
 Leader: 44798 (login)
   Seat: seat0; vc1
    TTY: tty1
 Remote: no
Service: login
   Type: tty
  Class: user
   Idle: no
   Unit: session-244.scope
         ├─44798 "login -- username"
         ├─45239 systemctl --user start --wait wayland-wm@labwc.desktop.service
         └─45535 waitpid -e 45660

Sep 28 13:22:17 hostname systemd[1]: Started session-244.scope - Session 244 of User username.
Sep 28 13:22:19 hostname uwsm[45239]: Default compositor ID unchanged: labwc.desktop.
Sep 28 13:22:19 hostname uwsm[45239]: Selected compositor ID: labwc.desktop
                                                Command Line: labwc
                                            Plugin/binary ID: labwc
                                       Initial Desktop Names: labwc:wlroots
                                                        Name: labwc
                                                 Description: A wayland stacking compositor
Sep 28 13:22:19 hostname uwsm[45239]: Units unchanged.
Sep 28 13:22:19 hostname uwsm[45239]: Starting labwc.desktop and waiting while it is running...
Sep 28 13:22:20 hostname uwsm[45535]: waitpid: Holding login session 244 open until PID 45660 exits
```
